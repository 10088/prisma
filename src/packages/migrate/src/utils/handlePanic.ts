import chalk from 'chalk'
import { RustPanic, sendPanic, link, isCi, logger } from '@prisma/sdk'
import prompt from 'prompts'
import { wouldYouLikeToCreateANewIssue } from './getGithubIssueUrl'

export async function handlePanic(
  error: RustPanic,
  cliVersion: string,
  binaryVersion: string,
): Promise<void> {
  if (isCi() && Boolean((prompt as any)._injected?.length) === false) {
    throw error
  }

  await panicDialog(error, cliVersion, binaryVersion)
}

async function panicDialog(error, cliVersion, binaryVersion) {
  const errorMessage = error.message
    .split('\n')
    .slice(0, Math.max(20, process.stdout.rows))
    .join('\n')

  logger.log(`${chalk.red('Oops, an unexpected error occured!')}
${chalk.red(errorMessage)}

${chalk.bold('Please help us improve Prisma by submitting an error report.')}
${chalk.bold(
  'Error reports never contain personal or other sensitive information.',
)}
${chalk.dim(`Learn more: ${link('https://pris.ly/d/telemetry')}`)}
`)

  const response = await prompt({
    type: 'select',
    name: 'value',
    message: 'Submit error report',
    initial: 0,
    choices: [
      {
        title: 'Yes',
        value: true,
        description: `Send error report once`,
      },
      {
        title: 'No',
        value: false,
        description: `Don't send error report`,
      },
    ],
  })

  const reportFailedMessage = `${chalk.bold.red(
    'Oops. We could not send the error report.',
  )}`

  if (response.value) {
    let reportId: number | void
    try {
      logger.log('Submitting...')
      reportId = await sendPanic(error, cliVersion, binaryVersion)
    } catch (error) {
      logger.log(reportFailedMessage)
    }

    if (reportId) {
      logger.log(
        `\n${chalk.bold(
          `We successfully received the error report id: ${reportId}`,
        )}`,
      )
      logger.log(`\n${chalk.bold('Thanks a lot for your help! üôè')}`)
    }
  }
  await wouldYouLikeToCreateANewIssue({
    prompt: !response.value,
    error,
    cliVersion,
    binaryVersion,
  })
}
