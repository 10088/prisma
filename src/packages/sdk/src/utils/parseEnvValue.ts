import { EnvValue } from '@prisma/generator-helper'
import chalk from 'chalk'

export function parseProviderEnvValue(object: EnvValue) {
  return parseEnvValue(object, 'provider')
}

export function parseBinaryTargetsEnvValue(object: EnvValue) {
  return parseEnvValue(object, 'binaryTargets')
}

/**
 * Parses the EnvValue and return the string value
 *
 * - If there is no env var just return the value
 * - If there is an env var it will be resolve and returned.
 * - If there is an env var is present but can't be resolved an error will be thrown
 */
function parseEnvValue(object: EnvValue, type: 'provider' | 'binaryTargets') {
  if (object.fromEnvVar && object.fromEnvVar !== null) {
    const value = process.env[object.fromEnvVar]
    if (!value) {
      throw new Error(
        `Attempted to load ${type} value using \`env(${
          object.fromEnvVar
        })\` but it was not present. Please insure that ${chalk.dim(
          object.fromEnvVar,
        )} is present in your Environment Variables`,
      )
    }
    return value
  }
  return object.value
}
