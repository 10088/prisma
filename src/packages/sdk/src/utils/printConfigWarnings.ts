import chalk from 'chalk'
import { logger } from '..'

export function printConfigWarnings(warnings: string[]) {
  if (warnings && warnings.length > 0) {
    const message = warnings
      .map((warning) => `${chalk.yellow('warn')} ${warning}`)
      .join('\n')
    logger.info(message)
  }
}
