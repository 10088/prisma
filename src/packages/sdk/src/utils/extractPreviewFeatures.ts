import { ConfigMetaFormat } from '../engine-commands'
import { parseProviderEnvValue } from './parseEnvValue'

export function extractPreviewFeatures(config: ConfigMetaFormat): string[] {
  return (
    config.generators.find(
      (g) => parseProviderEnvValue(g.provider) === 'prisma-client-js',
    )?.previewFeatures || []
  )
}
