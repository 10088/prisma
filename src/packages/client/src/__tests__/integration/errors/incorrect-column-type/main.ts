import { getTestClient } from '../../../../utils/getTestClient'

test('incorrect-column-type', async () => {
  const PrismaClient = await getTestClient()
  const prisma = new PrismaClient({
    log: [
      {
        emit: 'event',
        level: 'query',
      },
    ],
  })

  try {
    const data = await prisma.user.findMany()
  } catch (e) {
    prisma.$disconnect()
    expect(e.message).toMatchSnapshot()
  }
})
