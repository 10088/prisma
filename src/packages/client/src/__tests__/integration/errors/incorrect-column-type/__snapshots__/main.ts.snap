// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`incorrect-column-type 1`] = `
"
[31mInvalid [1m\`prisma.user.findMany()\`[22m invocation in[39m
[31m[4m/root/code/prisma/src/packages/client/src/__tests__/integration/errors/incorrect-column-type/main.ts:15:36[24m[39m

[2m  [90m11 [39m  ],[22m
[2m  [90m12 [39m})[22m
[2m  [90m13 [39m[22m
[2m  [90m14 [39mtry {[22m
[31m[1mâ†’[22m[39m [2m[90m15 [39m  const data = await prisma.user.findMany([22m
  Attempted to serialize scalar '123' with incompatible type 'String'"
`;
